zkevm-circuits/src/mpt_circuit/account_leaf/account_non_existing.rs

```diff
@@ -85,426 +85,102 @@ Note that the selector (being 1 in this case) at `s_main.rlp1` specifies whether
+Note 1: The hash of the only account is checked to be the state root.
+Note 2: There is no nil_object case checked in this gate, because it is covered in the gate
+above. That is because when there is a branch (with nil object) in the first level,
+it automatically means the account leaf is not in the first level.
+
 */
 
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, Default)]
 pub(crate) struct AccountNonExistingConfig<F> {
     _marker: PhantomData<F>,
 }
 
 impl<F: FieldExt> AccountNonExistingConfig<F> {
-    #[allow(clippy::too_many_arguments)]
     pub fn configure(
-        meta: &mut ConstraintSystem<F>,
-        q_enable: impl Fn(&mut VirtualCells<'_, F>) -> Expression<F> + Copy,
-        not_first_level: Column<Advice>,
-        s_main: MainCols<F>,
-        c_main: MainCols<F>,
-        accs: AccumulatorCols<F>,
-        sel1: Column<Advice>, /* should be the same as sel2 as both parallel proofs are the same
-                               * for non_existing_account_proof */
-        power_of_randomness: [Expression<F>; HASH_WIDTH],
-        fixed_table: [Column<Fixed>; 3],
-        address_rlc: Column<Advice>,
-        check_zeros: bool,
+        meta: &mut VirtualCells<'_, F>,
+        cb: &mut MPTConstraintBuilder<F>,
+        ctx: MPTContext<F>,
     ) -> Self {
-        let config = AccountNonExistingConfig {
+        let proof_type = ctx.proof_type;
+        let not_first_level = ctx.position_cols.not_first_level;
+        let accs = ctx.accumulators;
+        let address_rlc = ctx.address_rlc;
+
+        let rot_key_s = -ACCOUNT_NON_EXISTING_IND;
+        let rot_first_branch = -(ACCOUNT_NON_EXISTING_IND - 1 + BRANCH_ROWS_NUM);
+        let rot_branch_init = rot_first_branch - 1;
+
+        circuit!([meta, cb.base], {
+            let account = AccountLeafInfo::new(meta, ctx.clone(), rot_key_s);
+
+            // Make sure is_wrong_leaf is boolean
+            require!(account.is_wrong_leaf(meta, true) => bool);
+
+            ifx! {a!(proof_type.is_non_existing_account_proof) => {
+                ifx! {account.is_wrong_leaf(meta, true) => {
+                    // Note: currently, for non_existing_account proof S and C
+                    // proofs are the same, thus there is never a placeholder branch.
+                    let (key_rlc_prev, key_mult_prev, is_key_odd) = ifx! {a!(not_first_level) => {
+                        let branch = BranchNodeInfo::new(meta, ctx.clone(), true, rot_branch_init);
+                        (a!(accs.key.rlc, rot_first_branch), a!(accs.key.mult, rot_first_branch), branch.is_key_odd())
+                    } elsex {
+                        (0.expr(), 1.expr(), false.expr())
+                    }};
+                    // Calculate the key and check it's the address as requested in the lookup
+                    let key_rlc_wrong = key_rlc_prev + account.key_rlc(meta, &mut cb.base, key_mult_prev, is_key_odd, 1.expr());
+                    require!(a!(address_rlc) => key_rlc_wrong);
```

Comment:
This works as it should, but it confused me a little that you call `account.key_rlc` function where `account` stores data from the `ACCOUNT_LEAF_KEY_S` row whereas we need to compute the RLC for the data in the  `ACCOUNT_NON_EXISTING` row.

It works because `key_rlc` function uses rotation 0, not the actual `AccountLeafInfo` rotation, so `account.key_rlc` actually computes `account_wrong.key_rlc` (`account_wrong is defined a couple of lines below). Perhaps it could be changed so that `key_rlc` uses the actual rotation?